pkg/backup/backup_compaction.go

```diff
@@ -0,0 +1,613 @@
+		ctx, p.User(), p.ExecCfg(), incrementalURIs, collectionURIs, resolvedSubdir,
+	)
+	return
+}
+
+// maybeWriteBackupLock attempts to write a backup lock for the given jobID, if
+// it does not already exist. If another backup lock file for another job is
+// found, it will return an error.
+func maybeWriteBackupLock(
+	ctx context.Context,
+	execCtx sql.JobExecContext,
+	dest backupdest.ResolvedDestination,
+	jobID jobspb.JobID,
+) error {
+	foundLockFile, err := backupinfo.CheckForBackupLock(
+		ctx,
+		execCtx.ExecCfg(),
+		dest.DefaultURI,
+		jobID,
+		execCtx.User(),
+	)
+	if err != nil {
+		return err
+	}
+	if foundLockFile {
+		return nil
+	}
+	if err := backupinfo.CheckForPreviousBackup(
+		ctx,
+		execCtx.ExecCfg(),
+		dest.DefaultURI,
+		jobID,
+		execCtx.User(),
+	); err != nil {
+		return err
+	}
+	return backupinfo.WriteBackupLock(
+		ctx,
+		execCtx.ExecCfg(),
+		dest.DefaultURI,
+		jobID,
+		execCtx.User(),
+	)
+}
+
+// prepareCompactedBackupMeta prepares the manifest, job details,
+// and resolved destination for the compacted backup based on the chain of backups.
+func prepareCompactedBackupMeta(
+	ctx context.Context,
+	execCtx sql.JobExecContext,
+	jobID jobspb.JobID,
+	oldDetails jobspb.BackupDetails,
+	backupChain []backuppb.BackupManifest,
+) (
+	manifest *backuppb.BackupManifest,
+	details jobspb.BackupDetails,
+	dest backupdest.ResolvedDestination,
+	err error,
+) {
+	newDetails, err := makeCompactionBackupDetails(
+		oldDetails, backupChain[1:],
+	)
+	if err != nil {
+		return
+	}
+	dest, err = backupdest.ResolveDest(
+		ctx,
+		execCtx.User(),
+		newDetails.Destination,
+		newDetails.EndTime.AddDuration(10*time.Millisecond),
+		execCtx.ExecCfg(),
+	)
+	if err != nil {
+		return
+	}
+	if err = maybeWriteBackupLock(ctx, execCtx, dest, jobID); err != nil {
+		return
+	}
+
+	var tenantSpans []roachpb.Span
+	var tenantInfos []mtinfopb.TenantInfoWithUsage
+	insqlDB := execCtx.ExecCfg().InternalDB
+	if err = insqlDB.Txn(ctx, func(ctx context.Context, txn isql.Txn) error {
+		tenantSpans, tenantInfos, err = getTenantInfo(ctx, execCtx.ExecCfg().Codec, txn, newDetails)
+		return err
+	}); err != nil {
+		return
+	}
+	_, manifest, err = getBackupDetailAndManifest(
+		ctx, execCtx.ExecCfg(), tenantSpans, tenantInfos, newDetails, execCtx.User(), dest,
+	)
+	if err != nil {
+		return
+	}
+	// The manifest returned by getBackupDetailAndManifest have its
+	// start and end times set based on the job details. We need to update them to
+	// reflect the compacted backup's start and end times.
+	manifest.StartTime, manifest.EndTime = newDetails.StartTime, newDetails.EndTime
+
+	manifest.IntroducedSpans, err = compactIntroducedSpans(ctx, *manifest, backupChain)
+	if err != nil {
+		return
+	}
+	return
+}
+
+// concludeBackupCompaction completes the backup compaction process after the backup has been
+// completed by writing the manifest and associated metadata to the backup destination.
```

Comment:
maybe once this is jobified, this could become a helper shared by regular and compacted backups